traffic_ops/traffic_ops_golang/topology/topologies.go

```diff
@@ -150,6 +150,20 @@ func (topology *TOTopology) Validate() error {
 	rules["empty cachegroups"] = topology_validation.CheckForEmptyCacheGroups(topology.ReqInfo.Tx, cacheGroupIds, dsCDNs, false, nil)
 	rules["required capabilities"] = topology.validateDSRequiredCapabilities()
 
+	for _, cdnId := range dsCDNs {
+		err, dsId := dbhelpers.GetDSByCDNIdTopology(topology.ReqInfo.Tx.Tx, cdnId, topology.Name)
+		if err != nil {
+			return fmt.Errorf("failed to get a response from DB: %s", err)
+		}
+		for _, id := range dsId {
+			topoCachegroupNames := topology.getCachegroupNames()
+			userErr, sysErr, _ := dbhelpers.CheckOriginServerInCacheGroupTopology(topology.ReqInfo.Tx.Tx, id, topology.Name, topoCachegroupNames)
```

Comment:
Ok, I think I might've been wrong about being able to reuse this same `CheckOriginServerInCacheGroupTopology` function for this validation. When we update a topology, we don't want to make N DB queries (where N is the number of delivery services assigned to the topology). Preferably, we would make a single DB query that gets all the DS data we need, then we can do the validation from there.

Reply:
Done.